---
title: "Workshop 10 - Quality control"
output: 
  html_document: 
    theme: yeti
---

```{css, echo=FALSE}
.exercisebox {
  padding: 1em;
  background: #ffffcc;
  color: black;
  border: 2px #ffffcc;
  border-radius: 10px;
}

.examplebox {
  padding: 1em;
  background: #e6ffff;
  color: black;
  border: 2px #e6ffff;
  border-radius: 10px;
}

.center {
  text-align: center;

```

#  {.tabset}

## Part 1 - Intra-patient Contamination

### Intra-patient Contamination

![](https://raw.githubusercontent.com/lochlanf/GNA5012/main/Workshop%20Materials/Week_10/img/intrapatient.png)

In this section of the workshop we will examine intrasample contamination (usually referred to as tumour purity in the context of cancer). Tumour purity is a very important aspect of somatic variant calling. Low tumour purity can significantly reduce variant calling sensitivity and lead to us missing clinically important variants.

Understanding the purity of our tumour samples is an important consideration when interpretting the output of somatic variant calling. If you have sequenced a tumour with low purity and have relatively few variants in your filtered vcf, it is worth sequencing the library to a higher depth.

### Inferring tumour purity using computational approaches

There are several computational approaches to assessing tumour purity. These include methods that assess RNA, DNA methylation, and copy number profiles. One of the more popular tools for inferring tumour purity is [ASCAT (Allele Specific Copy number Analysis of Tumours)](https://github.com/VanLoo-lab/ascat)

::: exercisebox
**Exercise 1**

-   ASCAT is an R package. Start up Rstudio and create a new project called "Week 10". In the project directory, create data and results subdirectories (folders).

-   Download the workshop data from Moodle, extract the .zip file within the data directory

-   Open up R and create a new Rmarkdown file. This is where we will be entering our scripts. Review week 1 for instructions on how to do this if you are unsure.

-   Execute the following code in the **console**. This will install ASCAT and it's dependencies.

    ```{r, eval=FALSE}
    if (!require("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
    BiocManager::install(version = "3.15")
    BiocManager::install(c('GenomicRanges','IRanges'))
    install.packages("devtools")
    devtools::install_github('VanLoo-lab/ascat/ASCAT')
    library(ASCAT)
    ```

-   Please attract the attention of a tutor if you are not able to load the ASCAT library.
:::

There are several steps involved in somatic copy number, purity and ploidy estimation using ASCAT. These are:

```{r, eval = FALSE}
ascat.loadData()
ascat.correctLogR()
ascat.aspcf()
ascat.plotSegmentedData()
ascat.runAscat()
ascat.metrics()
```

::: exercisebox
**Exercise 2**

-   In small groups, discuss the purposes of each step. Refer to the manuscript, user guide and other online resources to help you with this task.
:::

### Running ASCAT

For this workshop, we have B-Allele Frequency and Log R data from 100 breast cancer and normal DNA samples (`Germline/Tumor_BAF.txt` and `Germline/Tumor_LogR.txt`). All patients are female. These are tab delimited text files that we can load in using the `ascat.loadData` function. In exercise three, we will load and process these data using ASCAT

::: exercisebox
**Exercise 3**

1.  Create a new code chunk. In the code chunk, use the `ascat.loadData` function to load the BAF and LogR files. Remember to assign them to a object using the `<-` operator.
    -   `ascatAnalysis <- ascat.loadData(..., gender = rep('XX', 100), genomeVersion = 'hg19')`
    -   Note, we have used the `rep` function to indicate that all samples have two X chromosomes as the patients are female. We have also set the genome version to hg19, which is the genome corresponding to the SNP positions in the file.
2.  Using the `ascat.CorrectLogR` function, correct your newly created object. You will need to specify three arguments in this command, the object name (ie. ascatAnalysis), the GC content file `GC_example.txt`, and the replication timing file `RT_example.txt`. Make sure you overwrite your previous object.
    -   `ascatAnalysis <- ascat.correctLogR(ascatAnalysis, GCcontentfile = '...', replictimingfile = '...')`
3.  Next, we need to use the `ascat.aspcf` function to denoise and segment the data by identifying change-points. The results of this will be written out to file, therefore we should provide the `out.dir` argument to direct these files to our results directory.
    -   `ascatAnalysis <- ascat.aspcf(ascatAnalysis, out.dir = "...")`
    -   **Note:** It is normal to get warnings at this point, they can be safely ignored.
4.  Before we run ASCAT, it is a good idea to plot our segmented data and do some sanity checks to make sure segmentation looks correct. We can use the `ascat.plotSegmentedData` function for this.
    -   `ascat.plotSegmentedData(ascatAnalysis, img.dir = "...")`
5.  We can now run ASCAT with the `ascat.runAscat` function. We want to retain segmental copy number data, we can do this by setting `write_segments` to `T`. Remember to use the `img.dir` argument to direct the output files to your results directory. You'll need to assign the output to a new variable.
    -   `ascatOutput <- ascat.runAscat(ascatAnalysis, write_segments = T, img.dir = "...")`
6.  Our final step is to collect QC metrics. Use the `ascat.metrics` function to gather these metrics
    -   `QC <- ascat.metrics(ascatAnalysis, ascatOutput)`
:::

```{r, eval = FALSE, echo=FALSE}
#if (!require("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install(version = "3.15")
#BiocManager::install(c('GenomicRanges','IRanges'))
#install.packages("devtools")
#devtools::install_github('VanLoo-lab/ascat/ASCAT')
library(ASCAT)

ascatAnalysis <- ascat.loadData(Tumor_LogR_file = "data/ASCAT/Tumor_LogR.txt", Tumor_BAF_file = "data/ASCAT/Tumor_BAF.txt", Germline_LogR_file = "data/ASCAT/Tumor_BAF.txt", Germline_BAF_file = "data/ASCAT/Germline_BAF.txt", gender = rep('XX',100), genomeVersion = "hg19") # isTargetedSeq=T for targeted sequencing data
ascatAnalysis <- ascat.correctLogR(ascatAnalysis, GCcontentfile = "data/ASCAT/GC_example.txt", replictimingfile = "data/ASCAT/RT_example.txt")
ascatAnalysis <- ascat.aspcf(ascatAnalysis, out.dir = "results/")
ascat.plotSegmentedData(ascatAnalysis, img.dir = "results/")
ascat.output <- ascat.runAscat(ascatAnalysis, write_segments = T, img.dir = "results/")
QC <- ascat.metrics(ascatAnalysis,ascat.output)
```

### Reviewing ASCAT Results

Let's review our ASCAT results. Note that there are files in your results file and data in the QC table. Let's review the QC table first. 

::: exercisebox
**Exercise 4**

-   Using the `str` function, identify the structure of the QC object. How many rows and columns does it have?
-   Using the `summary` function, generate summary statistics for the `purity` and `ploidy` columns. What is the average tumour purity?
-   Using ggplot2, generate a histogram of tumour purity.
-   Calculate the heterozygous VAF (1/2 of tumour purity) for the sample with the highest and lowest tumour purity. Cross-reference this with the image below. If we wish we have >80% SNV sensitivity in all samples, how deep should we sequence the libraries?
![](https://raw.githubusercontent.com/lochlanf/GNA5012/main/Workshop%20Materials/Week_10/img/IlluminaVAF.png)
-   We will review the sunrise plots as a class
:::


## Part 2 - Inter-species contamination

### Inter-species contamination

![](https://raw.githubusercontent.com/lochlanf/GNA5012/main/Workshop%20Materials/Week_10/img/interspecies.png)

Inter-species contamination refers to contamination of your sample with nucleic acids that are derived from another organism. We can use several techniques to identify this contamination bioinformatically. We can use k-mer matching based approaches to match k-mers to large databases, or we can use mapping based approaches to map subsets of our target reads to selected reference genomes. K-mer matching based approaches are fast and unbiased, however they require an exceptionally large amount of compute resources as they load the entire database into RAM. In this part of the workshop, we will use mapping based methods to identify contamination.

### Background

Dr Mick Ops works a laboratory that uses three model organisms, C.elegans, C.intestinalis, D.melanogaster and the broiler chicken. You have been given a dataset from the researcher. The data is 2x75bp and was sequenced on a NextSeq500. You notice that it is mapping very poorly. You are told that the data is from broiler chickens, but have your doubts. The files are in `week_10/data`.

### FastQ-Screen

FastQ-Screen is a mapping based program that can identify potential contaminators in your sequencing data. It uses an aligner (bowtie, bowtie2 or BWA) to map a subset of the reads to candidate genomes and reports hits on each genome. This can be very useful when dealing with a handful of known contaminating organisms. I have configured FastQ-Screen to use four reference genomes, each if which were indexed with `bwa index`. You will find these in your `ref` directory. 

::: exercisebox
**Exercise**

-   Using fastq_screen, identify possible contaminants

-   The syntax for this program is `fastq_screen OPTIONS forward_reads reverse_reads`, where OPTIONS refers to any custom flags and forward_reads and reverse_reads refers to the path to your sequencing reads

-   There are three flags that you will need to use

    -   -conf path_to_configuration_file

    -   \--aligner bwa

    -   \--outdir path_to_results_directory

-   This will generate a text file in your output directory with mapping rates to the four model organisms
:::

### Group Discussion

::: exercisebox
**Exercise**

-   Based on the data from the Fastq-Screen, what organism contributes the greatest proportion of sample?
-   Dr Mick Ops submitted his samples for library preparation in a 96 well plate. See the below plate diagram. Using your knowledge of the sampling, extraction and library preparation processes, outline where this error may have occured.
![](https://raw.githubusercontent.com/lochlanf/GNA5012/main/Workshop%20Materials/Week_10/img/plate.png)
:::

## Part 3 - DNA analysis - Putting it all together

### DNA sequencing and analysis - Putting it all together

This week ends our analysis of DNA sequencing data. We have

-   Performed QC on raw sequencing reads and assessed sequence quality using FastQC

-   Evaluated other quality control metrics and tools

-   Compared the role of different sequencing platforms in detecting variants

-   Aligned and preprocessed genomes

-   Called candidate germline variants using HaplotypeCaller

-   Evaluated genome-wide genotyping data and genome-wide association studies

-   Called small indels and structural variants using Strelka2 and Manta

-   Called somatic variants using MuTect2

Most samples that you interact with will be processed using similar workflows to the ones that you have been exposed to in class.

::: exercisebox
**Exercise 1**

-   In groups and on the whiteboards around the room, draw workflow diagrams describing how samples from a cancer patient go from DNA extraction to variant calls

-   Include the following in your diagrams:

    -   What samples you will sequence

    -   Quality control

    -   Alignment

    -   Germline variant calling

    -   Somatic variant calling

    -   Somatic Copy number analysis

    -   Structual variant calling

-   You should include all the tools, including their input and outputs.
:::

::: examplebox
**Example**

-   Here is an example of a flow chart that shows how I might process paired-end sequencing from raw sequencing reads to duplicate marked alignments. Your diagrams should look similar to this.

![](https://raw.githubusercontent.com/lochlanf/GNA5012/main/Workshop%20Materials/Week_10/img/Example_flow.jpg)
:::
